### DISCUSSION TITLE
Allow messages window cursor to automatically go to find instances and keep focus

### DISCUSSION THREAD

**Discussion by advice2020** [geany/geany#4026] [Ideas]
You can use the Messsages window to view instances of a keyword found when using "Find in files".
Currenly you have to use your mouse to click on each instance to view them, or you can use arrow keys up/down to move cursor but in order to view that item you have to hit Enter key which takes focus out of the Messages window.

Suggesting to allow user to be able to use the arrows keys to move the cursor/highlight up and down and have the results of that line automatically show in the text pane (the same as it does when you left click on them), but still keep the focus in the Messages window.
This would allow user the ability to navigate through these found items using just the arrow keys.  The Enter key action can remain which would then bring the focus to that item in the text pane.

If anyone knows if this is currently possible, please let me know, I feel this would be very helpful.

Thank You

EDIT:
Just realized that although the keyboard actions do not exist from this request, there is an option to generate find results for just the current document in the Ctrl+F find dialog.   If you expand the "Find All" drop down, there is "Mark" "In Session" and "In Document" which is the one that generates search results in the Messages Window as I was asking for.

**Comment by elextr** [comment]
Find in files searches files not open in Geany.  If up/down arrow showed each line that would mean every file would have to be opened as the arrow went past it, bad design.  Therefore enter will open the selected line _only_ and it is intended to move to the selected match, not just to show it, after all the line that has a match is shown in the message list.  This is the intended behaviour and won't change.

If any other behaviour is to be added it will need to use different keybindings.

**Comment by eht16** [comment]
If you press Space on a selected line in the messages window, the corresponding file is opened but the focus will stay on the messages window so you can still navigate with the cursor keys.

I think this is what you requested.

**Reply by eht16** [reply]
@advice2020 And there is yet another possibility:

You could use the "Next message" and "Previous message" keybindings to navigate in the search results.

**Comment by elextr** [comment]
Probably s/b documented since (AFAIK) its not a default thing but added at [here](https://github.com/geany/geany/blob/eb831d7c846abdf567c79fe0e0d64ce212acf7ca/src/msgwindow.c#L177).

**Comment by eht16** [comment]
This feature exists because it was already requested many, many years ago:
https://sourceforge.net/p/geany/feature-requests/245/ and implemented in 
https://github.com/geany/geany/commit/334e7092ea40bc1cf33b7c6bac8dcccfa827bbd7.

**Comment by elextr** [comment]
> This feature exists

Fine, but...

> Probably s/b documented 

so users can find out without reading the code ;-)

**Reply by eht16** [reply]
Actually, I didn't read the code before, I just tried it.

Adding this to the docs is a good idea. Where could we add this?

**Reply by elextr** [reply]
I looked at the manual and was about to make a simple PR when I got confused, so ran out of time, so questions.

1. its confusing that there is a message window (although its actually a pane not a window, but can live with that) and then inside it there is a message tab, and I got confused which of the 68 uses of "message" in the document refers to the "window" and when it refers to the "tab" or refers to some which else since most of the 68 do not distinguish
2. good refer to "find usage" and "find in all files" at [here](https://github.com/geany/geany/blob/3122fb1dbcc1906df32a89f0a9d19253fa44288f/doc/geany.txt?plain=1#L244) but then it says "other actions" which is not described so who knows whats shown
3. "find usage" has a note its results [are in window tab](https://github.com/geany/geany/blob/3122fb1dbcc1906df32a89f0a9d19253fa44288f/doc/geany.txt?plain=1#L1329) :heavy_check_mark: 
4. but "find in files" just below has no mention where its results are shown :x: 
5. but it is mentioned in the "Find" dialog description results of some searches
6. and some discussion of "message" is really about compiler results ... I think??
7. keybindings that relate to the message tab contents are changable, but the enter and space are hard coded and so don't get mentioned in keybindings
8. other than the keybindings there is no description anywhere about the message window message tab message lines format from a number of unmentioned contents.  Mouse or up/down arrow selection of message lines are in keybinding, but nothing about enter to switch the source on the message line to edit focus, or space to switch the edit but no focus switch.
9. maybe?? a new section for the message window message tab with:
   - a list of all the actions as well as find usage and find all files
   - all actions that put stuff in the message window message tab should refer to this new section
   - explanation of the message location
   - a list of the keybindings and hard coded functions (assuming they are all the same for all those actions that use it)

And that point I give up.

Edit: a thought, probably it would be better as "Miscellaneous Window" not "Message Window" since it has tabs "Status" "Compiler" "Message" which are sort of messages, but also "Scribble" and "Terminal" neither being messages.

**Comment by advice2020** [comment]
Thank you guys for the responses.
At least there are some options that can be used.
Glad to see this generated some activity.

@eht16
Thank you for your responses.
Not as convenient as just using arrows, but Space Bar works well because it does keep the focus in the messages window which is what I was looking for, only difference is have to hit it each time, thank you for mentioning this.
The next/previous message keybindings mentioned, that works even betters because functions more like what I was looking for.  I had overlooked those, thank you for mentioning these as well.

@elextr 
I agree with what you are saying about the terminology used, it is quite confusing.
Agree not a window, it is docked in the bottom bar.
Agree name is confusing because as you mentioned contains much more than messages.
Maybe can name it something like "Bottom Bar" (like how Geany uses "Side Bar") or "Bottom Panel"
Also to me, confusing that search results end up in what is called messages tab,  I mean it works and makes use of the messages tab, but to me would not expect it to end up in that tab.
I think it is a good idea to give this its own section in the manual, because as you mentioned there is a lot of information about it not shown in the manual currently.

Also
I have updated my original thread post above because although Geany does not have the exact keyboard actions I was looking for, it does actually have this same feature already as was mentioned above.
When using Ctrl+F to open Find dialog, if you expand the "Find All" drop down, where the "Mark" button is located, there is also "In Session" and "In Document" which generates the search results in the Messages Window.
When generate find results this way, it would be nice if the arrows up/down could be used because no conflict would occur because only one file, but I would not add because then would be confusing because would not be consistant with the other uses of the Messages tab.

Anyways, if I have some time, I will try to part some of these items mentioned in these responses into separate request posts just in case future developers are interested in making some updates.

Thank you guys again
